joomla/joomla-cms
# [4.0] Set up console application

Partial Pull Request for Issue #17879

### Summary of Changes

The Framework's Console package is introduced into the CMS with this pull request and the base scaffolding for its use wired up.

The main changes are pulling in the package and its dependencies (this includes a couple other Symfony components), building the base console application for the CMS environment, and migrating the core CLI scripts to be console commands.  Note the Smart Search CLI is purposefully not migrated as this is a component specific script and we should handle this one when we have reached a point where components are supported with this structure.

Of note, the console application will import two plugin groups to be able to handle system events (those dispatched from the console application itself as well as events in other parts of the system that may be loaded).  One is the core `system` group, the other is a new `console` group (this in essence allows installing plugins that would only be used in the console application).

### Testing Instructions

With this PR applied in your local environment, from the command line run `php cli/joomla.php`.  You should see an output similar to the below:

```sh
Joomla! 4.0.0-dev

Usage:
  command [options] [arguments]

Options:
  -h, --help            Display the help information
  -q, --quiet           Flag indicating that all output should be silenced
      --ansi            Force ANSI output
      --no-ansi         Disable ANSI output
  -n, --no-interaction  Flag to disable interacting with the user
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Available commands:
  help                            Show the help for a command
  list                            List the application's available commands
 cache
  cache:clean                     Cleans expired cache entries
 update
  update:extensions:check         Checks for pending extension updates
  update:joomla:remove-old-files  Removes old system files
```

Without any arguments or options, you get the list of available commands by default.  Right now there are five commands; the two defaults from the Framework Console package (`help` and `list`), and the three commands from the CMS.

Running `php cli/joomla.php help <command>` gives a detailed help window for the specific command, i.e. this for the `cache:clean` command.

```sh
Usage:
  cache:clean

Options:
  -h, --help            Display the help information
  -q, --quiet           Flag indicating that all output should be silenced
      --ansi            Force ANSI output
      --no-ansi         Disable ANSI output
  -n, --no-interaction  Flag to disable interacting with the user
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
  The cache:clean command cleans the system cache of expired entries
  
  php cli/joomla.php cache:clean
```

With the CMS correctly installed, running any command should just work without issue.
